content/docs/reference-react-dom-server.md

```diff
@@ -40,47 +40,89 @@ The following methods can be used in the environments that don't support streams
-    onShellError(error) {
-      // Something errored before we could complete the shell so we emit an alternative shell.
-      res.statusCode = 500;
-      res.send(
-        '<!doctype html><p>Loading...</p><script src="clientrender.js"></script>'
-      );
+    onError(err) {
+      // TODO: when is this called?
+      // ??? Called if one of the suspended content blocks failed to render.
+      didError = true;
+      console.error(err);
     },
-    onAllReady() {
-      // If you don't want streaming, use this instead of onShellReady.
-      // This will fire after the entire page content is ready.
-      // You can use this for crawlers or static generation.
+  }
+);
+```
 
-      // res.statusCode = didError ? 500 : 200;
-      // res.setHeader('Content-type', 'text/html');
-      // stream.pipe(res);
-    },
+#### Opting-out of inline scripts
+
+If you’d like to support clients that can’t run client-side JavaScript such as crawlers, you can wait until **all** suspended blocks are ready and then stream the HTML in one go.
+
+To do so, implement the `onAllReady` callback and start piping the stream into your response here. The `onError` option will be called if an error occurs during rendering, while may mean you HTML will only be partially outputted up to the point the error occurred.
+
+```javascript
+let didError = false;
+const stream = renderToPipeableStream(
+  <App />,
+  {
     onError(err) {
+      // Called if an error occurred while rendering.
       didError = true;
       console.error(err);
     },
+    onAllReady() {
+      // If you don't want streaming, use this instead of onShellReady.
+      // This will fire after the entire page content is ready.
+      // You can use this for crawlers or static generation.
+      res.statusCode = didError ? 500 : 200;
+      res.setHeader('Content-type', 'text/html');
+      stream.pipe(res);
+    },
   }
 );
 ```
 
-See the [full list of options](https://github.com/facebook/react/blob/14c2be8dac2d5482fda8a0906a31d239df8551fc/packages/react-dom/src/server/ReactDOMFizzServerNode.js#L36-L46).
+#### Hydrating to add interactivity {#rendertopipeablestream-hydrate}
+
+If you call [`ReactDOM.hydrateRoot()`](/docs/react-dom-client.html#hydrateroot) on a DOM node that already has this server-rendered markup, React will preserve it and only attach event handlers, allowing you to have a very performant first-load experience.
+
+#### Options
+
+- `onShellReady?: () => void` — called when an initial shell is ready to be streamed. Start piping into your response here and suspended content blocks will "pop in" via inline <script> tags.
+- `onShellError?: () => void` — called if an error occurs while rendering the initial shell.
+- `onAllReady?: () => void` — use instead of `onShellReady` to wait for all suspense boundaries to be ready before streaming. Start piping into your response here and the entire HTML tree will be outputted in one go.
+- `onError?: (error: mixed) => void` — (called when?)
```

Comment:
I’m not sure when this is called and how it behaves different using `onShellReady` vs `onAllReady`